src/distributions/utils.rs

```diff
@@ -131,34 +131,159 @@ macro_rules! simd_float_from_int {
 /// implement it as a trait so we can write the same code for `f32` and `f64`.
 /// Only the comparison functions we need are implemented.
 pub trait CompareAll {
+    type Mask;
     fn all_lt(self, other: Self) -> bool;
     fn all_le(self, other: Self) -> bool;
+    fn all_finite(self) -> bool;
+    fn finite_mask(self) -> Self::Mask;
+    fn gt_mask(self, other: Self) -> Self::Mask;
+    fn ge_mask(self, other: Self) -> Self::Mask;
+
+    // Decrease all lanes where the mask is `true` to the next lower value
+    // representable by the floating-point type.
+    fn decrease_masked(self, mask: Self::Mask) -> Self;
+}
+
+// Implement functions available in std builds but missing from core primitives
+#[cfg(not(std))]
+pub(crate) trait Float : Sized {
+    type Bits;
+
+    fn is_nan(self) -> bool;
+    fn is_infinite(self) -> bool;
+    fn is_finite(self) -> bool;
+    fn to_bits(self) -> Self::Bits;
+    fn from_bits(v: Self::Bits) -> Self;
+}
+
+// Implement functions on f32/f64 to give them APIs similar to SIMD types
+pub(crate) trait FloatAsSIMD : Sized {
+    #[inline(always)]
+    fn lanes() -> usize { 1 }
+    #[inline(always)]
+    fn splat(scalar: Self) -> Self { scalar }
+    #[inline(always)]
+    fn extract(self, index: usize) -> Self { assert_eq!(index, 0); self }
+    #[inline(always)]
+    fn replace(self, index: usize, new_value: Self) -> Self { assert_eq!(index, 0); new_value }
```

Comment:
I think it would be more appropriate to use `debug_assert` here?

Reply:
I was trying to replicate the behavior of these function on real SIMD types, and hoping that the compiler would optimize the checks away.

Ultimately doesn't really matter right now. These functions are only used in tests.

Happy to go either way.

I did change `decrease_masked` though to not cast the bool flag to an int. Instead I made it part of the contract that at least one "lane" must be set and added `debug_assert`s to ensure that. That way we can always subtract `1`. This might have gotten optimized away, but we might as well be safe.